server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java

```diff
@@ -0,0 +1,146 @@
+package org.elasticsearch.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class RemoteConnectionManager {
+
+    private final String clusterAlias;
+    private final ConnectionManager connectionManager;
+    private final AtomicLong counter = new AtomicLong();
+    private volatile List<Transport.Connection> connections = Collections.emptyList();
+
+    RemoteConnectionManager(String clusterAlias, ConnectionManager connectionManager) {
+        this.clusterAlias = clusterAlias;
+        this.connectionManager = connectionManager;
+    }
+
+    public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfile,
+                              ConnectionManager.ConnectionValidator connectionValidator,
+                              ActionListener<Void> listener) throws ConnectTransportException {
+        connectionManager.connectToNode(node, connectionProfile, connectionValidator, new ActionListener<>() {
+            @Override
+            public void onResponse(Void aVoid) {
+                try {
+                    // TODO: Propagate the connection to the listener
+                    Transport.Connection newConnection = connectionManager.getConnection(node);
+                    addConnection(newConnection);
```

Comment:
If possible, I think we should purely rely on the TransportConnectionListener events to add and remove nodes, and not manually recreate that logic. In particular, the logic as it stands here has a problem (see comment below in `removeConnection`). If need be, we can add for example extend the `TransportConnectionListener.onNodeConnected` method to also contain the `Connection` object as parameter.

Note that this also means that we need strong guarantees from the TransportConnectionListener, the order in which its methods are called (I don't think we have that today) and should also have additional tests on that.